msbuild/Xamarin.MacDev.Tasks.Core/Tasks/ComputeBundleLocationTaskBase.cs

```diff
@@ -0,0 +1,333 @@
+			// Contrary to the "TargetPath" and "Link" metadata, this value doesn't specify the filename itself, only the containing directory name.
+			var destinationSubDirectory = item.GetMetadata ("DestinationSubDirectory");
+			if (!string.IsNullOrEmpty (destinationSubDirectory))
+				virtualPath = Path.Combine (destinationSubDirectory, virtualPath);
+
+			return virtualPath;
+		}
+
+		void ReportUnknownPublishFolderType (ITaskItem item)
+		{
+			var publishFolderType = item.GetMetadata ("PublishFolderType");
+
+			var metadata = item.GetMetadata ("CopyToOutputDirectory");
+			if (!string.IsNullOrEmpty (metadata)) {
+				Log.LogWarning (MSBStrings.E7090 /* The 'PublishFolderType' metadata value '{0}' on the item '{1}' is not recognized. The file will not be copied to the app bundle. If the file is not supposed to be copied to the app bundle, remove the '{2}' metadata on the item. */, publishFolderType, item.ItemSpec, "CopyToOutputDirectory");
+				return;
+			}
+
+			metadata = item.GetMetadata ("CopyToPublishDirectory");
+			if (!string.IsNullOrEmpty (metadata)) {
+				Log.LogWarning (MSBStrings.E7090 /* The 'PublishFolderType' metadata value '{0}' on the item '{1}' is not recognized. The file will not be copied to the app bundle. If the file is not supposed to be copied to the app bundle, remove the '{2}' metadata on the item. */, publishFolderType, item.ItemSpec, "CopyToPublishDirectory");
+				return;
+			}
+
+			Log.LogWarning (MSBStrings.E7088 /* The 'PublishFolderType' metadata value '{0}' on the item '{1}' is not recognized. The file will not be copied to the app bundle. */, publishFolderType, item.ItemSpec);
+		}
+
+		// 'item' is not supposed to have a PublishFolderType set
+		PublishFolderType ComputePublishFolderType (IList<ITaskItem> items, ITaskItem item)
+		{
+			var filename = item.ItemSpec;
+			var targetPath = item.GetMetadata ("TargetPath");
+			if (!string.IsNullOrEmpty (targetPath))
+				filename = Path.Combine (Path.GetDirectoryName (filename), Path.GetFileName (targetPath));
+
+			// Check if the item came from @(BundleResource), @(Content) or @(EmbeddedResource)
+			if (resourceFilesSet.Contains (Path.GetFullPath (item.ItemSpec)))
+				return PublishFolderType.Resource;
+
+			// Assemblies and their related files
+			if (filename.EndsWith (".dll", StringComparison.OrdinalIgnoreCase)) {
+				return PublishFolderType.Assembly;
+			} else if (filename.EndsWith (".exe", StringComparison.OrdinalIgnoreCase)) {
+				return PublishFolderType.Assembly;
+			} else if (filename.EndsWith (".pdb", StringComparison.OrdinalIgnoreCase)) {
+				return PublishFolderType.Assembly;
+			} else if (filename.EndsWith (".dll.mdb", StringComparison.OrdinalIgnoreCase) || filename.EndsWith (".exe.mdb", StringComparison.OrdinalIgnoreCase)) {
+				return PublishFolderType.Assembly;
+			} else if (filename.EndsWith (".config", StringComparison.OrdinalIgnoreCase)) {
+				return PublishFolderType.Assembly;
+			}
```

Comment:
The all return 'PublishFolderType.Assembly', could we have done this with a loop through he known extensions:
```suggestion
			foreach(var extension in new string [] {".dll", ".exe",".pdb", "dll.mdb", ".exe.mdb", ".config"}) {
				if (filename.EndsWith (extension, StringComparison.OrdinalIgnoreCase))
					return PublishFolderType.Assembly;
                        }
```

Or you can create a helper function that returns if it is an assembly checking the extension and do a simple if.